modules/templates/handler.rb

```diff
@@ -7,23 +7,28 @@ module Proxy::Templates
   class Handler < ::Proxy::HttpRequest::ForemanRequest
     include Proxy::Log
 
-    def get_template(kind, token, static = false)
-      opts = { :token => token,
-               :url   => Proxy::Templates::Plugin.settings.template_url
-      }
-      opts[:static] = static if static
-      request = request_factory.create_get("/unattended/#{kind}", opts)
-      logger.debug "retrieving a template from %s/%s" % [Proxy::Templates::Plugin.settings.template_url, request.path]
-      res = send_request(request)
+    def get_template(kind, request, params)
+      url = Proxy::Templates::Plugin.settings.template_url
+      proxy_ip = URI.parse(url).host
+      opts = params.clone.merge(:url => url)
+      opts.delete("kind")
+      opts.delete("splat")
+      opts.delete("captures")
+      proxy_headers = Hash[request.env.select{|k,v| k =~ /^HTTP_/ && k !~ /^HTTP_(VERSION|HOST)$/}.map{|k,v| [k[5..-1],v]}].merge("X-Forwarded-For" => "#{request.ip}, #{proxy_ip}")
+      proxy_req = request_factory.create_get("/unattended/#{kind}", opts, proxy_headers)
+      logger.debug "Retrieving a template from %s/%s" % [url, proxy_req.path]
+      logger.debug "HTTP headers: #{proxy_headers.inspect}"
+      res = send_request(proxy_req)
 
       # You get a 201 from the 'built' URL
-      raise "Error retrieving #{kind} for #{token} from #{uri.host}: #{res.class}" unless ["200", "201"].include?(res.code)
+      raise "Error retrieving #{kind} for #{opts.inspect} from #{uri.host}: #{res.class}" unless ["200", "201"].include?(res.code)
+      Proxy::Log.logger.info "Template: request for #{kind} using #{opts.inspect} at #{uri.host}"
       res.body
     end
 
-    def self.get_template kind, token, static = false
+    def self.get_template(kind, request, params)
```

Comment:
Hmmm I am not so sure about this one. Objects should be passed instead of it's data. I prefer nicer code than nicer tests.